coach/src/user/model.coffee

```diff
@@ -46,29 +46,41 @@ User =
     course = _.findWhere @_course_data, ecosystem_book_uuid: collectionUUID
     course and _.detect(course.roles, (role) -> role.type is 'teacher')
 
-  status: (collectionUUID) ->
-    course = @getCourse(collectionUUID)
+  status: (collectionUUID, enrollmentCode) ->
+    course = @getCourse(collectionUUID, enrollmentCode)
+
     isLoggedIn: @isLoggedIn()
     isLoaded:   @isLoaded
     isRegistered: !!course?.isRegistered()
     preValidate: (not @isLoggedIn()) and (not course?.isValidated())
 
-  isEnrolled: (collectionUUID) ->
-    Boolean(@isLoggedIn() and @getCourse(collectionUUID))
+  isEnrolled: (collectionUUID, enrollmentCode) ->
+    Boolean(@isLoggedIn() and @getCourse(collectionUUID, enrollmentCode))
+
+  getCourse: (collectionUUID, enrollmentCode, options) ->
+    initialFilter = _.extend(ecosystem_book_uuid: collectionUUID, options)
+    filterForEcosystem = _.partial(_.where, _, initialFilter)
+    sortyByJoinedAt = _.partial(_.sortBy, _, (course) -> course.getRole()?.latest_enrollment_at or '')
+    filters = [filterForEcosystem]
+
+    if enrollmentCode
+      filterForEnrollmentCode = _.partial(_.filter, _, (course) ->
+        course.enrollment_code is enrollmentCode or
+          _.find(course.periods, enrollment_code: enrollmentCode)
+      )
+      filters.push(filterForEnrollmentCode)
 
-  getCourse: (collectionUUID) ->
-    _.chain(@courses)
-      .where( ecosystem_book_uuid: collectionUUID )
-      .sortBy( (course) -> course.getRole()?.latest_enrollment_at or '' )
-      .last()
-      .value()
+    filters.push(sortyByJoinedAt)
+
+    _.last(_.compose(filters...)(@courses))
```

Comment:
Nice way to compose the search! I'd suggest calling them `operations` rather than `filters` though since it also sorts.